core/app/c/[communitySlug]/pubs/[pubId]/page.tsx

```diff
@@ -95,66 +115,78 @@ export default async function Page(props: {
 
 	// We don't pass the userId here because we want to include related pubs regardless of authorization
 	// This is safe because we've already explicitly checked authorization for the root pub
-	const pubPromise = tryCatch(
-		getPubsWithRelatedValues(
-			{ pubId: params.pubId, communityId: community.id },
-			{
-				withPubType: true,
-				withRelatedPubs: true,
-				withStage: true,
-				withStageActionInstances: true,
-				withMembers: true,
-				depth: 3,
-			}
-		)
-	);
+	const pubPromise = getPubsWithRelatedValuesCached(pubId, community.id);
 
 	const actionsPromise = getStageActions({ pubId: pubId }).execute();
 
+	// sadly two steps
+	const [pub, availableViewForms, availableUpdateForms] = await Promise.all([
+		pubPromise,
+		getAuthorizedViewForms(user.id, pubId).execute(),
+		getAuthorizedUpdateForms(user.id, pubId).execute(),
+	]);
+
+	if (!pub) {
+		notFound();
+	}
+
+	if (!availableViewForms.length) {
+		redirect(`/c/${params.communitySlug}/unauthorized`);
+	}
+
+	const hasAccessToCurrentForm = !availableViewForms.find(
+		(form) => form.slug === formSlug || (form.isDefault && !formSlug)
+	);
+	const defaultForm = availableViewForms.find((form) => form.isDefault);
+	if (!hasAccessToCurrentForm) {
+		const redirectQuery = defaultForm ? "" : `?form=${availableViewForms[0].slug}`;
+
+		// redirect to first available form
+		redirect(`/c/${communitySlug}/pubs/${pub.id}/edit${redirectQuery}`);
+	}
```

Comment:
similar to the `edit` page, redirect users to the correct form if they can't access the current one